### ISSUE TITLE
13 - Cat Images

### ISSUE THREAD

**Issue by TheBeege** [YUHEE1984/CatsOneday#13]
# Goal

You wanted to see images of your cats, right?

# Context

The timing of this project is perfect. With the dawn of generative AI, we can now generate images of your cats. This does introduce some major complication. AI images aren't generated instantly - they take a few seconds. If your API clients wait for several seconds for a response, there is a high chance they could "timeout," or hit some threshold where they decide something likely failed and stop waiting. To solve this, we'll create a job system.

What's a job system? It's where clients request that work be started, then they can repeatedly ask for the status of the work. Once the work is completed, they can request the result of the work. This kind of system is perfect for long-running tasks, like generating an AI image. A system like this uses what we call "asynchronous programming." This is writing code that doesn't necessarily run in exactly the order you define it. It's not synchronized, or "a-sync" ("a" being "not" and "sync" being "synchronized"). 

We'll also go over how to customize your API inputs and outputs, how to do one-to-many relationships, how to call an external API, how to do application configuration, and more. Beware: this will be the hardest issue, but it will also be the most rewarding.

Let's get started!

# Technical Requirements

A new section?! It's getting real, now. Think you can handle it? Good.

Given some cat, a user will want to generate an image for that cat. This means that image generation will always be tied to a cat. In this case, it makes sense for images to only be within the context of a cat. In REST, this would mean that endpoints like `/cat/{id}/images` would work well.

Generating images takes some time, as described above. When a user requests a new image to be generated, we don't want to make them wait for the image to be produced. On requests for new images, we should create a new image generation job. Once that job is created, we should provide an endpoint for users to see the status of that job. When the job is complete, users will need some way to retrieve the image generated by that job.

When requesting images, users won't want just images of their cat. They'll want to see their cat doing some action in some location in some lighting according to some art style. Our endpoint to request images should allow users to provide this information.

Also, note that the image request will be a text-based prompt. We won't want users to enter the prompt themselves. We should generate this prompt on-demand based on the information the user entered. We should store the specific inputs the user provided but provide the prompt we generate from it as the request.

When users view their cats, they'll always want to see the available images. We should include references to the images for users to fetch whenever they're ready to look at them.

Finally, images and other binary data should be stored on some kind of "blob storage" or file storage. If we were using the cloud, this would be something like AWS S3. In our case, we'll just use a local directory tos tore the images in.

Here's a summary of the endpoints, their inputs, and their outputs:

* `POST /cats/{id}/imageRequests`
    * Inputs
        * `action` - a string describing what the cat is doing
        * `location` - a string describing where the cat is
        * `lighting` - a string describing what the lighting conditions are
        * `artStyle` - a string describing the art style that a user wants their cat rendered in
    * Outputs
        * `imageRequestId` - an ID representing the job to generate the image
        * `createdAt` - a timestamp of when the request was made
        * All the fields provided in the call to create the request
    * Notes
        * Calling this endpoint **should not** wait for the image generation request to complete. It should return the job ID to the user immediately after the request is sent off.
* `GET /cats/{catId}/imageRequests/{imageRequestId}`
    * Outputs
        * `id` - the ID of the image request
        * `status` - a string representing the status of the image generation job. This should be one of `CREATED`, `ERROR`, or `COMPLETED`.
        * `errorReason` - a string describing the reason for the error. When the `status` is not `ERROR`, this should be `null`.
        * `createdAt` - a timestamp representing when the request was made
        * `prompt` - the prompt used to generate the image
    * Notes
        * Make sure to check that the image request belongs to the cat in the request. We don't want users spying on other users' cats!
* Update! `GET /cats/{id}`
    * Outputs
        * Add an `images` attribute to the `Cat` object that is an array.
            * This array should include include the ID of the request, the status of the request, the timestamp when the request was created, the prompt to generate the request, and if the request is completed, the URL for retrieving the image generated.
* Update! `GET /cats`
    * Outputs
        * Add a `lastImageUrl` attribute to the `Cat` object that gives just the URL of the last generated image. If there's no completed image, then `lastImage` should be `null`.

The section you've just read would be typical of requirements provided to you by a senior engineer or a technical product manager. If you aren't provided with documentation like the above, it's likely safe for you to request this. Once you become a more mid-level engineer, you would be expected to come up with the endpoint definitions on your own given just the descriptions in the first part of this section. As a senior, you would be writing all of this.

Before you start on the steps below, try to plan out how you would build this without help.

* How would you create the new endpoints?
* How would change your inputs and outputs?
* What data would you need to store?
* What logic would you need to add?

You're not going to know the answer to all of these. Try to search around online for tips on how to accomplish these various goals. Remember: a key skill of an engineer is to take larger problems they don't understand and break them down into smaller, better understood problems. Another key skill is to ask questions. Once you've taken some time to think about how to solve these and maybe search around for different ways of doing things, then you can move on to the steps below. Good luck!

# Steps

## Defining Our Interface

When creating a new feature, it's best to start with the interface, so you'll know what data you'll need and how you'll need to manipulate that data. We'll start with the API endpoints. After that, we'll build out the data layer. Lastly, we'll build out the service layer to do all the complex logic. Let's go!

### `POST /cats/{id}/imageRequests`

1. Open up your `CatController` class, and create a new method `generateImage`. This method will return nothing (`void`) for now.
2. Add a `@PostMapping("/{catId}/imageRequests")` annotation on the method.
3. Add a parameter `Long catId` to get the path parameter. Don't forget to add the `@PathVariable` annotation.
5. Create a new `CreateImageRequestDto` class in the `dto` package.
6. Add the `@Data` and `@NoArgsConstructor` annotations on the class.
7. Create the following `private String` attributes: `action`, `location`, `lighting`, `artStyle`.
    * We'll use this class to define the format of our request body for creating our image request.
8. Back in your `CatController`, add a parameter to the `generateImage` method named `requestInputs` as your new `CreateImageRequestDto` type, and make sure to use the `@RequestBody` annotation on that parameter.

That's a good enough start for now. This will define our input for our endpoint. Since our output will include the ID of the request along with the timestamp and input data, we can likely use the model for the output. So we can leave this as it is now. Let's move forward!

### `GET /cats/{catId}/imageRequests/{imageRequestId}`

1. Still in the `CatController` class, let's add a `getImageRequest` method. It should take in a `catId` parameter and an `imageRequestId` parameter. Both should be `Long` types, and they should both have the `@PathVariable` annotation. Leave the return type as `void` for now.
2. Add a `@GetMapping("/{catId}/imageRequests/{imageRequestId}")` annotation on the method.
    * By this point, you should be starting to predict how we're doing some parts here and there. If not, go review some past endpoints you've created, and think more before writing the next ones.
3. In your `dto` package, create a new class `ImageRequestWithPromptDto`. It should have a `Long id`, a `String errorReason`, a `String prompt`, and an `Instant createdAt`. All of the attributes should be `private`, as usual.
4. Add the `@Data` and `@NoArgsConstructor` annotations to the class.
5. I hope you noticed that `status` is missing. We're going to do something special for that, since it must be one of a fixed set of values. Create a new file `ImageRequestStatus` in your `model` package. Instead of making it a `class`, make it a `public enum`.
6. Inside your `enum` definition, just add these separated by commas: `CREATED`, `ERROR`, and `COMPLETED`.
7. Go back to your `ImageRequestWithPromptDto` class and add a `status` attribute with the `ImageRequestStatus` type. Don't forget to make it `private`.
8. Finally, go back to your `CatController`, and set the output type of the `getImageRequest` method to your new `ImageRequestWithPromptDto` class. Your editor may complain about the return type not being correct, so just drop a `return null` in the method body for now.

So far, so good! Just two more. Let's make good use of your new knowledge of DTOs.

### Update! `GET /cats/{id}`

1. We'll start with the inner `images` objects from the technical requirements for this endpoint. In your `dto` package, create a new `ImageRequestWithPromptAndUrlDto` class.
2. Add the `id`, `status`, `createdAt`, `prompt`, and `url` attributes. You can reference the `ImageRequestWithPromptDto` class for the types. Set the `url` attribute's type to `String`. 
5. In your `CatInListDto` class, a new `images` attribute, which will be a `Collection` of the `ImageRequestWithPromptAndUrlDto` type. The `java.util` package has the `Collection` class you want.
    * Collection is an interface that handles any list-like or array-like objects. It's useful when you don't really care exactly how items are collected, just that they're in a collection.

We're not going to populate this new `images` attribute, yet. We'll do that soon. One remains!

### Update! `GET /cats`

1. Create a new `CatWithLastImageDto` class in your `dto` package.
2. You can copy the `CatWithImagesDto` class and replace `images` with `lastImageUrl`. Make sure to set the type to `String`.
4. In the `CatController`'s `get` method, change `return catService.find(name)` to `return catService.find(name).stream().map(CatWithLastImageDto::new).toList()`. Don't worry about what this is or how it works. It's way beyond your level for now. Remember, it's only temporary!
5. Run `./gradlew bootRun` and check [http://localhost:8080/swagger-ui/index.html](http://localhost:8080/swagger-ui/index.html).
6. Look through each endpoint that you created or changed, and make sure it matches the technical requirements section. Note that we haven't set the proper return type for `POST /cats/{id}/imageRequests` yet, so don't worry about that one.

Excellently done! You've got the API definition mostly complete! You can now use this to help guide how you create your database layer and business logic layer.

## Handling Job Data

1. Just like our `Cat` model class, let's create a class for `ImageGenerationJob` in our `model` package.
2. Again like our `Cat` model, add these annotations to the class: `@Entity`, `@NoArgsConstructor`, `@AllArgsConstructor`, and `@Data`.
3. Let's add an `id` attribute, just like our `Cat` model. I'll let you find the proper scoping, annotations, and type to use, based on the `Cat` model.
4. Create a few `String` attributes to handle the user's inputs for the request: `action`, `location`, `lighting`, `artStyle`.
5. Create a `status` attribute as a `ImageRequestStatus`. This will store the status of the request to the image generation system.
6. Create an `errorReason` attribute as a `String`. This will store the error message if the request somehow fails.
7. Create a `path` attribute as a `String`. This will store where the image from the job is located.
8. Create a `createdAt` attribute as a `Instant` type. Add a `@CreatedDate` annotation right above it. Usually, you'll want some field with a `@CreatedDate` annotation and field with a `@LastModifiedDate` annotation on every entity to help with debugging and auditing. We've been lazy up to now, but having the created date for these images is useful for sorting and such.
9. Create a `cat` attribute with the type `Cat`. Add a `@ManyToOne` annotation above this attribute. Also add a `@JoinColumn(name = "cat_id", nullable=false)` annotation above it. This will help connect the image to its cat.
10. Now, go to the `Cat` class. Add a new `images` property with the type `Set<ImageGenerationJob>`. Note that the `Set` should be the `Set` in the `java.util` package, not Hibernate's `Set`. Give it the annotation `@OneToMany(mappedBy = "cat")`. In this annotation, the `mappedBy = "cat"` part is telling Spring Boot which attribute in the `ImageGenerationJob` model should store the `Cat` entity. This will finish the relationship between `Cat` and `ImageGenerationJob`s.
11. In the `Cat` class's `handleBreeding` method, don't forget to update the call to create a `new Cat`. You just need to add an extra `null` as the last parameter, since there is no set of images yet.
12. Run your server so that Spring updates the database with your changes. You can use `./gradlew bootRun` or your IDE's run function.
13. Visit your adminer UI at [http://localhost:8000/](http://localhost:8000/). Login (check `compose.yml`) and make sure that the `cat` table and `image_generation_job` table look like they have all the right fields.

Great job! You've laid the groundwork for a job system. This will allow us to track the jobs to create new AI images of our cats. 

## Setting Up Async

Credit to [Baeldung on how to set up async](https://www.baeldung.com/spring-async). If my explanation below isn't clear, try theirs.

1. In the `org.yuhee.catsoneday` package, create a new class `SpringAsyncConfig`. We'll use this to enable async for our project.
2. Add two annotations to the class: `@Configuration` and `@EnableAsync`.

That's it! There's more we could do if needed, but this should be sufficient. Async is normally a very complicated thing, but for basic configurations, Spring Boot makes it easy.

## Set up Stability API

Often when using an outside API, you'll be provided with an API "key" to identify yourself for using the API. This acts as both a username and password in a single string. We'll set up a secure way to store API keys and other secrets, then we'll sign up for the Stability API and store our API key.

1. Sign up for the Stability AI API platform: [https://platform.stability.ai/docs/getting-started](https://platform.stability.ai/docs/getting-started)
    * Note that signing up using a Google account gets you 25 credits for free. This should be enough for you to make sure things work if you don't just request images for fun.
    * If you want to request a lot of images, you may want to pay a little bit of money. Don't worry. The pricing is pretty good.
2. After signing up, click your profile image / account info in the top right corner.
3. You'll see a list of API keys. Click the copy icon (looks like one square behind another) to copy it.
4. In your project's `src/main/resources` folder, create a new file: `application.sensitive.properties`.
    * `.properties` files are used to store configurations for our application, like we did in 12 for saving the cats to the database. We're using a separate file here because we don't want our secrets saved with our other configurations.
5. In this file, write `stability-api-key=<my key>` and replace `<my key>` with the API key you copied from the Stability website.
6. In the root of your project, open `.gitignore`. Add `application.sensitive.properties` anywhere in the file. I usually add my custom entries at the top of the file.
    * What does this do? `.gitignore` tells Git to ignore files matching these names. We don't want your API showing up in Github! Someone could use your API key to spend all of your Stability API credits. By adding this file to `.gitignore`, we reduce the chance of accidentally publishing this secret information.

## Requesting the AI Image

1. Create a new interface `ImageRequestService` in your `service` package.
3. Create a new method called `requestImage` that takes in a `CreateImageRequestDto` object and a `Long` object that will be the `Cat`'s ID. It should output an `ImageRequestWithPromptDto` object. It should also throw a `CatNotFoundException`.
4. Create a new class `ImageRequestServiceImpl` in your `service` package. Make it implement the `ImageRequestService` interface. Make sure to create the required method.
5. Add the `@Service` annotation to the class.
6. Add a new private property as a String: `stabilityApiKey`.
7. Create a new constructor for `ImageRequestServiceImpl`. Take in a String for the Stability API key as one of the parameters. Before that parameter, add this annotation: `@Value("${stability-api-key}")`. This is similar to the `@PathVariable` annotation we use in our controller classes.
    * This is how we can inject data from `.properties` files. Spring Boot will see this annotation and search through the `.properties` files for the `stability-api-key` setting. When it finds the setting, it will store the value in our property, `stabilityApiKey`.
8. Create another `private` property to store a `CatRepository` object. In your constructor, add another parameter for a `CatRepository` and set the property to this parameter.
9. In your `requestImage()` method, create a new `Optional<Cat>` object, and set its value to the output of your `CatRepository`'s `findById()` method, providing the `Long` `Cat`'s ID parameter. 
10. If your `Optional<Cat>` object is empty, `throw` a new `CatNotFoundException`, providing the `Cat`'s ID.
11. After your check for empty, create a new `Cat` object and set it to the value you `get()` from the `Optional<Cat>` object.
12. Create a new `private` method called `generatePrompt` that outputs a `String`. It should take in a `CreateImageRequestDto` object and a `Cat` object.
13. In your `generatePrompt` method, return a string that the Stable Diffusion AI model will use to generate the image. Use the data from the `CreateImageRequestDto` object and the `Cat` object to fill out the text.
    * Having trouble thinking of something? You could try `"A <sex> cat <action>s at <location> in <lighting> lighting. The cat's fur is <color> colored with <furPattern> patterned fur. It has <eyeColor> eyes and weighs <weightInKg>kg. The art style is <artStyle>"`. Note that `<text>` is not the correct way to interpolate variables into text in Java. I'll let you figure that out on your own.
    * Later, if you want to get fancy, consider replacing the `generatePrompt` method with something using the [Builder design pattern](https://refactoring.guru/design-patterns/builder) ([Korean](https://refactoring.guru/ko/design-patterns/builder)) .
14. In the `requestImage` method, create a variable to store the prompt text, and call `generatePrompt` to populate that variable.
15. Create a new `HttpClient` variable, and initialize it with `HttpClient.newHttpClient();`.
16. Next, we'll use a [Builder pattern](https://refactoring.guru/design-patterns/builder) ([Korean](https://refactoring.guru/ko/design-patterns/builder)) to construct our HTTP request. Create a new `HttpRequest` object, and initialize it with `HttpRequest.newBuilder()`. Your IDE may show an error. Don't worry, yet. We'll fix it soon.
17. Next, we're going to chain some methods together. What is method chaining? It's when you call a method on the output of another method. It's very common with builders. It looks like `BuilderClass.newBuilder().setValue("").setAnotherValue(1).build();`. Note that you can add new lines before each period `.` to make it easier to read. On the `newBuilder()` method, chain another method `uri()`. In this `uri()` method, pass a new URI using `URI.create("https://api.stability.ai/v1/generation/stable-diffusion-xl-1024-v1-0/text-to-image")`. This is the address we can use to request an AI image. That `stable-diffusion-xl-1024-v1-0` bit is the model version, and it's the cheapest model available for Stable Diffusion. Later, we could make it configurable and/or use a nicer model, but let's keep things cheap while we're testing.
18. Let's chain another method after the `uri()` method. Add a call to a `header()` method. Pass `"Authorization"` as the first parameter and your `stabilityApiKey` as the second parameter. This will handle our login to the Stability API.
19. Let's chain another `header()` method, this time with `"Accept"` and `"image/png"` as the parameters. This tells the Stability API that we want a PNG image as the response.
20.  Add a `POST()` method call to our method chain. This tells the builder that we're making an HTTP request using the `POST` HTTP method. You may remember this from creating your first HTTP server. As a parameter to this `POST()` method, pass `HttpRequest.BodyPublishers.ofString("{\"text_prompts\": [{\"text\": \"" + prompt + "\", \"weight\": 0.5}]}")`, assuming the String variable storing your prompt is named `prompt`. This sends the prompt information as the request body of the HTTP request. This looks gross because we're writing JSON in a normal String. It would be better to use `jackson-databind`, but that's a little too complicated for now. I recommend reading more about it later.
21. Finally, chain one final method, `build()`, to our method chain. Now, the error should go away, and we're finished preparing the HTTP request.
22. We don't want our server to wait until the image is finished being generated to respond to the client, so we should handle receiving the AI image separately. Java has the idea of a `Future`, where code executes "asynchronously," or outside of the normal, sequential flow of the program. Instead of waiting for our HTTP request to complete, we'll accept a `Future` and handle processing the response in a separate thread. You can read about threads elsewhere. For ease, paste the below line, then we'll explain it:
    ```Java
    CompletableFuture<HttpResponse<byte[]>> futureResponse = client.sendAsync(httpRequest, HttpResponse.BodyHandlers.ofByteArray());
    ```
    * We're assming that your `HttpClient` object is stored in a variable named `client`. Update the above to match whatever variable name you used.
    * The `CompletableFuture` is the `Future` we talked about previously. It's a generic, so we must specify that the `CompletableFuture` will handle an `HttpResponse`. The `HttpResponse` is also a generic, so we must tell it that we expect an array of bytes as the response. Why an array of bytes? It's a reasonable way to represent image data.
    * The `HttpClient`'s `sendAsync()` method handles making the HTTP request and asynchronously receiving the response. It takes two parameters. The first is the `HttpRequest` that we built in the previous steps, and the second is how we should process the response. Since we're expecting an image back, we're choosing to use a byte array to handle the image with.
23. Now, let's create a new `ImageGenerationJob` object and store it in some variable. Use the properties of the `CreateImageRequestDto` and `Cat` to fill in the corresponding fields in `ImageGenerationJob`. Set the `status` as `ImageRequestStatus.Created`. Leave the `id`, `path`, and `createdAt` fields as `null` for now.
24. We'll need to store our image request jobs in the database, so we need a repository. In your `org.yuhee.catsoneday.repository` package, create a new `ImageGenerationJobRepository` interface. Make sure it extends `JpaRepository<ImageGenerationJob, Long>`.
25. Back in our `ImageRequestServiceImpl` class, we need to add a new `private ImageGenerationJobRepository` property.
26. Next, add another parameter to the `ImageRequestServiceImpl` class's constructor. This parameter should be an `ImageGenerationJobRepository` and should be stored in the class's property that you created in the previous step.
27. Back at the end of the `requestImage` method, let's use our `ImageGenerationJobRespository` object to `save()` the new `ImageGenerationJob`. Make sure to store the output in a variable! You can safely replace the original variable.
28. Let's step away from our `ImageRequestServiceImpl` class for just a moment and create a class to handle the HTTP response in a separate thread. In the `org.yuhee.catsoneday.service` package, create a new interface named `AiApiResponseHandler`. It should extend the `Runnable` interface.
29. In this interface, define a new method: `registerRequest()`. This should return nothing. It should take in an `ImageGenerationJob` and a `CompletableFuture<HttpResponse<byte[]>>` object representing our future image buffer response (hint: name it `futureImageBufferResponse`).
30. Now, create a new class: `AiApiResponseHandlerImpl`. It should implement the `AiApiResponseHandler` interface, and it should be annotated with the `@Service` annotation.
31. Create properties for three types of objects: `ImageGenerationJob`, `ImageGenerationJobRepository`, and `CompletableFuture<HttpResponse<byte[]>>`.
32. Create the `registerRequest()` method with the parameters as defined in the interface. Set the remaining class properties to these parameters.
33. Create a constructor that takes in an `ImageGenerationJobRepository` parameter and a `String` parameter that will be the directory path to store the cat image files.
34. Add Spring's `@Value()` annotation to the `String` parameter, and set the parameter to the value parameter as `"image-file-path"`.
35. For a brief moment, go to your `src/main/resources/application.properties` file. Add `image-file-path=images` and save it. Okay, back to `AiApiResponseHandlerImpl`.
29. In this `AiApiResponseHandlerImpl`, define a new `public `method: `run`. It returns nothing and has no parameters.
30. Let's leave the `AiApiResponseHandlerImpl` for now. We'll fill it in later. Go back to your `ImageRequestServiceImpl` class. 
31. Add a new property to store our `AiApiResponseHandler` and update our constructor to take in a parameter for it and assign the parameter to the property.
32. Back in the `requestImage` method, after running `save()` for our job, call the `AiApiResponseHandler` object's `registerRequest` method and provide the `ImageGenerationJob` and `CompletableFuture<HttpResponse<byte[]>>` objects as inputs.
31. Let's create a new thread that will run our `AiApiResponseHandlerImpl` in the background. Assuming you named the class property for it as `aiApiResponseHandler`, add the below to the end of your `requestImage()` method but before the `return null;`:
    ```java
    (new Thread(aiApiResponseHandler)).start()
    ```
    * We have some extra parenthesis `()` to make sure we call the `start()` method on the newly created thread. This thread will run in the background, allowing the rest of the `requestImage()` method to continue running.
31. Create a new variable of the `ImageRequestWithPromptDto` type. 
32. Using the `modelMapper`, map the job object resulting from your `save()` method to the `ImageRequestWithPromptDto` class.
33. On your `ImageRequestWithPromptDto` object, set the prompt to your `prompt` variable created earlier.
34. Finally, return your `ImageRequestWIthPromptDto` object.
35. Let's go to your `CatController` class and hook that up. First, let's add an `ImageRequestService` property.
37. Next, go to your `generateImage()` method. Change the output type from `void` to `ImageRequestWithPromptDto`.
38. Create a new variable to store an `ImageRequestWithPromptDto` object.
36. Next, `try` to call your `imageRequestService`'s `requestImage()` method, passing in the `Cat`'s ID and the request inputs. `catch` the `CatNotFoundException`. Assign the output of your `requestImage()` method to the `ImageRequestWithPromptDto` variable that we created in the previous step. 
37. Similarly to the `getOne()` method, `throw` a `ResponseStatusException`, providing the `HttpStatus.NOT_FOUND` parameter.
38. Lastly, `return` the `ImageRequestWithPromptDto` object.
39. Let's go back to your `AiApiResponseHandler` in your `org.yuhee.catsoneday.service` package.
40. Add a constructor that takes in three parameters: a `ImageGenerationJob` object, an `ImageGenerationJobRepository` object, and a `CompletableFuture<HttpResponse<byte[]>>` object.
    * The `ImageGenerationJob` object will store the model object that we'll need to update once the image is finished being generated.
    * We'll use the `ImageGenerationJobRepository` to update the `ImageGenerationJob` in the database.
    * The `CompletableFuture<HttpResponse<byte[]>>` will contain the image data once the request completes.
41. In the `AiApiResponseHandler` class, create properties to store the `ImageGenerationJob` and `CompletableFuture<HttpResponse<byte[]>>` objects. Update the constructor to set these properties.
42. Go back to your `ImageRequestServiceImpl`, and update the line where you create the new `AiApiResponseHandler` and the `Thread`. Pass in the image generation job, repository, and completable future objects as parameters to the constructor call.
43. Back in the `AiApiResponseHandler`, go to the `run()` method. Create a new `HttpResponse<byte[]>` to store the image data once we get it.
44. Create a `try ... catch` block.
45. In the `try` block, call the `CompletableFuture<HttpResponse<byte[]>>` class property's `get()` method, and store the output in the `HttpResponse<byte[]>` variable you created in the last step.
46. In your `catch` block, we want to catch all of the possible exceptions that `get()` can throw, so put `ExecutionException | InterruptedException | CancellationException exception` as the catch target.
47. In your `catch` block, set the `ImageGenerationJob`'s `status` to `ImageRequestStatus.ERROR`.
48. Also in the `catch` block, set the `ImageGenerationJob`'s `errorReason` to some descriptive text that hides what's happening internally on our server, like `"Failed to create AI image."`. 
49. Next in the `catch` block, use the `ImageGenerationJobRepository` to `save()` the `ImageGenerationJob` object.
49. Next in the `catch` block, print the output of `exception`'s `getMessage()` method.
50. Lastly in the `catch` block, `return`. This method runs outside of the normal code path, in the background. There's nowhere to report the error, but we should end the method's execution. Using `return` will simply end the method here.
50. After the `catch` block, we'll want to save the image data to a file.
51. Create a new `Path` variable from the `java.nio.file` package. We'll use this to write our retrieved image out to a file for later retrieval.
52. When writing the image to a file, we'll need a unique file name in order to avoid overwriting existing files or encountering other issues. Create a new `String` variable to store the file name. Assign its value to `UUID.randomUUID().toString()`.
    * `UUID` is short for universally unique identifier. UUIDs guarantee uniqueness but are quite large. For file names, this is suitable, as the length of the name doesn't matter, we need uniqueness, and the name doesn't need to carry any real meaning. The `ImageGenerationJob` has all of the context around the image file.
53. Create another `String` variable to store the full image file path. This string should start with your image files path from the method parameter, then have a slash (`/`), then the UUID file name, and finally the `.png` extension.
    * We know it's a `.png` file because the Stability API says it outputs `.png` files. If it were any other system, we would need to check the type of file being worked with.
52. Create a new `try ... catch` block.
    * Yes, we're doing lots of try/catch action. Anytime you deal with I/O (input and output) like network requests or files, there are many things that can go wrong. Always make sure to catch these exceptions and track things clearly. Ideally, you should log any exceptions clearly, too, but we'll cover that in another issue.
53. In the `try` block, call `Paths.get()`, passing the full image file path as the parameter. Assign this new object to your `Path` variable you created a few steps ago.
54. For the `catch` block, set a `InvalidPathException` as the target.
55. In the `catch` block, set the `ImageGenerationJob` object's `status` to `ImageRequestStatus.ERROR`. Set the `errorReason` to a meaningful message that hides the server implementation, like `"Failed to create valid file path."`. `save()` the job using the `ImageGenerationJobRepository`. Then print out the exception's `getMessage()` output for later logging. Lastly, `return`.
56. After the `catch` block, we need one last `try ... catch` block.
57. In the `try` block, call `Files.write()`. First, pass in the `Path` variable for the image file. As the second parameter, pass in the `HttpResponse<byte[]>` object you created towards the beginning of the method. You'll see there's an error. We need to get the bytes from the `HttpResponse`. Inside `write()`'s parameter, call the `HttpResponse<byte[]>` object's `body()` method. This will retrieve the bytes and pass them as the parameter for writing to the file. Lastly, pass in `StandardOpenOption.CREATE`. This tells Java that we intend to create this file.
59. The target of the `catch` block should be an `IOException`.
60. In the `catch` block, do the same thing we've been doing. Set the `ImageGenerationJob`'s status to `ImageRequestStatus.ERROR`. Set the error reason to something descriptive to the user but hiding the server internals, like `"Failed to save image file."`. Use the `ImageGenerationJobRepository` to `save()` the object. Finally, print out the exception's `getMessage()` output.
61. Now that we've handled all the possible errors and successfully retrieved and written the image to a file, let's store the file path in the job and report success. On the `ImageGenerationJob` object, set the `path` to the full image path. 
62. Set the `ImageGenerationJob` object's `status` to `ImageRequestStatus.COMPLETED`.
63. Using the `ImageGenerationJobRepository`, save the job. We're done with this method!
64. Next, we need to set up our server to actually serve the images we generate. Right now, they're stored on the server but don't have a way to show those files to users. In your `org.yuhee.catsoneday.controller` package, create a new class: `AiImageController`.
65. Add the `@RestController` annotation on the class.
66. Create a new `String` property to reference the image file path.
67. Create a constructor for the class, taking in a `String` for the image file path. I called mine `imagePathRoot` to prevent name collisions later. Add the `@Value()` annotation, passing in `"image-file-path"` as the parameter. This ensures that our `AiApiResponseHandler` and our `AiImageController` are using the same location to work with the image files.
68. In the constructor, set the parameter to the class property.
69. Create a new method: `getImage()`. It should output a `ResponseEntity<Resource>`. Note that the `Resource` should be from `org.springframework.core.io`. This is how we return a file. It should take in a `String` for the file name, named `fileName`, annotated with the `@PathVariable` annotation.
70. For the `getImage()` method, add a `@GetMapping()` annotation. Pass in the string `"images/{fileName}"` as the parameter.
71. In the method body, create a new `Path` variable, again from the `java.nio.file` package.
72. Create a `try ... catch` block. File I/O, remember? Lots of try/catch.
73. In the `try` block, we're going to do a little bit of method chaining again. On the `imagePathRoot`, call `resolve()`, passing in the `fileName` as the parameter. On that method call, chain a call to `normalize()`. Set the output of all this to the `Path` variable you created a couple of steps ago.
    * `resolve()` adds the given parameter as a folder to the path. Why do we do this as a method instead of just string manipulation? Well, operating systems are different. Unix systems like Mac and Linux use forward slash (`/`) as path separators, while Windows uses backslash (`\`) as path separators. The `resolve()` method handles this kind of logic for us.
    * `normalize()` basically cleans up the path. It removes unnecessary `./` parts and just generally makes the path as simple as possible.
74. For the `catch` block, set the target as an `InvalidPathException`.
75. Inside the `catch` block, print out a helpful message for later, like `"Unable to find requested file: " + fileName`.
76. Inside the `catch` block, `return` a method chain around `ResponseEntity`. `ResponseEntity` is a builder, so it will look familiar: `ResponseEntity.notFound().build()`. This will return a 404 not found response.
77. After the `try ... catch` block, we need to do some security. We need to make sure users aren't trying to access random other files on our server. Add a condition. The `Path` object for our file file path has a `startsWith()` method. The condition should check if the file path **does not** start with the `imagePathRoot` we defined. If not, that means they're trying to snoop around on our server.
78. If that condition is true, print out a message describing the final path they were trying to access and the `fileName` `String` that they provided to try to access that path. This would be useful for debugging and trying to protect from future attacks.
79. Still inside the condition, return a 404 not found response, similar to a few steps ago. The attacker doesn't need to know that we caught them.
80. After our condition, create a new `Resource` variable.
81. Create a new `try ... catch` block. This is our last one.
82. Inside the `try` block, create a new `UrlResource` object. For the constructor, pass in the output of our file path's `toUri()` method.
83. Set a `MalformedURLException` as the target for the `catch` block.
84. In the `catch` block, print out a helpful URL for later use, like `"Failed to build URL for file name: " + fileName`.
85. In the `catch` block, return a 404 not found response, like we've done before.
86. Return a different type of `ResponseEntity` chain, assuming you named your `Resource` variable as `resource`: `ResponseEntity.ok().body(resource)`. This will respond to the caller with the image resource read from the file. 
87. Finally, create an `images` directory in the root of your project folder. You're all done!
